### PR TITLE
Posterior sample

### PR DESCRIPTION
**Pull request by zouyuxin**
@gaow Could you help me to check the `compute_posterior_matrices_common_cov_R` function? Z is a P x J matrix. I use the for loop for j to do the sampling. The commented lines are the one using `sapply`. I am not sure which one is better. Also, I don't know if there is any method to avoid the for loop over j.

### CODE REVIEW

**Comment by gaow** [comment]
@zouyuxin thank you for the update. I think `sapply` should in principle at least make the code a lot clearer (if not faster), but I do not think I understood the necessity of those `if ... else` logic over `p`. Would you remind me what's special about `p=1`, the null?

> Also, I don't know if there is any method to avoid the for loop over j.

You mean to sample from matrix normal? Did you look for existing implementations?

**Comment by zouyuxin** [comment]
@gaow The special is not about the null. I want to set those rows to the samples. So I need to find the correct index (from which row to which row). One way is using `(Z_cumsum[p-1,j]+1):Z_cumsum[p,j]`. When p = 1, the first `Z_cumsum[1,j]` rows are the samples, so I need `1:Z_cumsum[1,j]` (the Z_cumsum[0,j] is numeric(0)). This is why I use the `if ... else` logic over p. I don't know if my explanation is clear. Do u have some suggestions about this?

When the data has equal sd, the `common_cov` version is more efficient than the general version, because we avoid the `for` loop over J. In my code, there is still a `for` loop over J. I don't know if there is a clearer way to avoid the loop.

**Comment by gaow** [comment]
@pcarbo Yuxin and I have simplified together her code to address to points raised in discussions above. I think I'm okay now with the general framework and core implementations. But there surely are coding practices etc that would be good to get your opinions on. Also please feel free to raise your concerns on more fundamental stuff if you spot anything wrong.

**Comment by pcarbo** [comment]
@gaow Thanks, if it okay with you, I will wait until you & Yuxin have addressed the issues with the unit tests before I review the code.

**Comment by gaow** [comment]
@pcarbo oh okay there are a couple of things I need to catch up with mash and susie hopefully the next few days. Will sort them out and pin you here again.

**Comment by gaow** [comment]
@pcarbo unit tests have been fixed. I also installed the versioning hook on my local clone so each commit now bumps internal version number, runs documentation and unit tests, before actually making a commit. We can now review & discuss this patch. I'll make some comments on the code of what I'm not sure about most (and pin you there).

File: R/mash.R [hunk 1]
```diff
@@ -36,6 +38,8 @@ mash = function(data,
                 add.mem.profile = FALSE,
                 algorithm.version = c("Rcpp","R"),
                 pi_thresh = 1e-10,
+                posterior_samples = 0,
```
Comment (gaow) [inline R/mash.R]: @pcarbo am not sure if I like this interface -- having `posterior_samples` and `seed` here; most awkwardly, `seed`. In general i'm still trying to get used to the "workhorse" function based software design which might seem natural to you, thus not a concern here.

Comment (pcarbo) [reply R/mash.R]: @gaow Thanks for checking in about this. Don't worry about it for now. It was a suggestion only.

**Pull request merged**